revenue_app/presto_connection.py

```diff
@@ -0,0 +1,62 @@
+import pandas as pd
+from pyhive import presto
+from pyhive.exc import DatabaseError
+
+
+class PrestoError(Exception):
+    pass
+
+
+def read_sql(file_name):
+    fd = open('{}.sql'.format(file_name), 'r')
+    sql_file = fd.read()
+    fd.close
+    return sql_file
+
+
+def query_presto(start_date, end_date, okta_username, okta_password, query):
+    connection = presto.connect(
+        'presto-tableau.prod.dataf.eb',
+        8443,
+        okta_username,
+        password=okta_password,
+        protocol='https',
+    )
+    dataframe = pd.read_sql(query.format(start_date, end_date), connection)
+    connection.close()
+    return dataframe
+
+
+def make_query(start_date, end_date, okta_username, okta_password, query_name):
+    try:
+        query = read_sql(query_name)
+        dataframe = query_presto(start_date, end_date, okta_username, okta_password, query)
+        return dataframe
+    except pd.io.sql.DatabaseError as exception:
+        error = exception.args[0].split('\n\n')[1]
+        if 'NewConnectionError' in error:
+            # Raised when not connected to the VPN
+            message = error.split('>:')[1].strip()
+            message += '.<br>Check you are connected to the VPN.'
+        elif 'SSLCertValidationError' in error:
+            # Raised when certificate not in env var
+            message = error.split("(1, '")[1].split("')))")[0].strip()
+            message += '.<br>Update your certificate '
+            message += '(<a href="https://docs.evbhome.com/intro/self_signed_certs.html">link</a>).'
+        elif 'Invalid credentials' in error:
+            # Raised when wrong okta username/password
+            message = error.split('<pre>')[1].split('</pre>')[0].strip()
+            message += '.<br>Check your okta username and password.'
+        else:
+            message = 'Unknown error.<br>'
+            message += error
+        raise PrestoError(message)
+    except DatabaseError as exception:
+        # Raised when you don't have permissions to a specific table
+        if query_name == 'transactions':
+            # this 'if' should dissapear once we know what to do with the transactions query
+            dataframe = pd.read_csv('datasets/transactions.csv')
+            return dataframe
+        message = exception.args[0]['message']
+        raise PrestoError(message)
```

Comment:
If all of this is gonna be handled by the exact same exception (should it?), we could move this logic to the Exception itself, so that the exception is the one that knows which message to display.

Reply:
Actually something weird is happening. `pandas.read_sql` catches other exceptions (`NewConnection`, `SSL`, `Invalid Credentials`) and throws its own exception (`DatabaseError`). I haven't found a way to avoid this.
It would be nice to move the login to the exception, but how can I do that?

Reply:
Even so, All that logic you do to show the correct message it's weirdly placed. I'm not certain it belongs in the `make_query` function.

Reply:
May be a `handle_presto_exception(exception)` method being called in the `except`?